Activate: Add support for internal remote feature flag

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8998 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR introduces the ability to use internal remote feature flags instead of relying on Firebase. The same mechanism is already used in the WordPress and Jetpack apps.

Since this work has architectural implications, I'm seeking as much feedback as possible to ensure that it is intuitive, user-friendly, and future-proof. Here are some initial considerations:

1. Use the existing `FeatureFlagStore` in FluxC.
2. Create a new enum class (`RemoteFeatureFlag`) that functions similarly to the existing `FeatureFlag` class, with a use case class `CheckRemoteFeatureFlagEnabled` for doing the checking.
3. Provide the ability to fetch and update all remote flags at once, including fetching by app version.
4. Provide the ability to get the value of a single remote flag from the database.

### Use cases:

#### Adding a new remote feature flag into the app
1. Update `RemoteFeatureFlag` enum with  a new constant containing the remote key, then
2. Update `CheckRemoteFeatureFlagEnabled` to handle the new enum constant. `PackageUtils.isDebugBuild()` can be used to force-enable a feature during development purposes, and can be removed as needed.

#### Fetching latest remote feature flags
`WPComRemoteFeatureFlagRepository.fetchFeatureFlags()` is available for this. This function behind-the-scene saves the fetched values to the DB, which is then used for checking individual remote flag value as below.

#### Checking whether a remote feature flag is enabled
Assuming `fetchFeatureFlags()` above has been called recently, inject and use the `CheckRemoteFeatureFlagEnabled(REMOTE_FEATURE_FLAG_CONSTANT)` usecase class.


### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

For now, ensure unit tests run properly. Do further testing once this functionality is actually used.


### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

hafizrahman committed:
Add: Starter content for class WPComRemoteFeatureFlagRepository with fetch functionality.

hafizrahman committed:
Add: Result and error handling for fetching feature flags.

hafizrahman committed:
Add: A way to get a single RemoteFeatureFlag based on a key.

hafizrahman committed:
Add: RemoteFeatureFlag class for setting up remote feature flags.

hafizrahman committed:
Detekt fix.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/config/WPComRemoteFeatureFlagRepository.kt:
```diff
@@ -0,0 +1,52 @@
+        }
+    }
+
+    fun getFeatureFlag(key: String): RemoteFeatureFlag? {
+        val flag = featureFlagsStore.getFeatureFlags().find { it.key == key }
+        return flag?.let { RemoteFeatureFlag(key = it.key, value = it.value) }
+    }
```
Nice job @hafizrahman. Other than the feedback Ondrej already shared over Slack which I agree with, I also have a couple suggestions 🙂

Let's assume you add the proposed changes from Slack thread:
```kotlin
    fun isFeatureEnabled(ff: RemoteFeatureFlag): Boolean {
        return when (ff) {
            RemoteFeatureFlag.LOCAL_NOTIFICATION_STORE_CREATION_READY,
            RemoteFeatureFlag.LOCAL_NOTIFICATION_NUDGE_FREE_TRIAL_AFTER_1D ->
                PackageUtils.isDebugBuild() || getFeatureFlag(ff.getRemoteKey())?.value ?: false
            else -> true
        }
    }
```

Then, do you need `fun getFeatureFlag` function at all? You could actually have something like: 
```kotlin
    fun isFeatureEnabled(flag: RemoteFeatureFlag): Boolean {
        return when (flag) {
            RemoteFeatureFlag.LOCAL_NOTIFICATION_STORE_CREATION_READY,
            RemoteFeatureFlag.LOCAL_NOTIFICATION_NUDGE_FREE_TRIAL_AFTER_1D ->
                PackageUtils.isDebugBuild() ||
                    featureFlagsStore.getFeatureFlags().find { it.key == flag.remoteKey }?.value ?: false
            else -> true
        }
    }
```

With that, you could also get rid of: 
```kotlin
 class RemoteFeatureFlag(
        val key: String,
        val value: Boolean
    )
```
I mean, what is the purpose of that model? (btw should it be a `data class`). Imho with 2 functions you should be good to go. One for refreshing/fetching the feature flags status from the API, which would correspond to what you already have in: `suspend fun fetchFeatureFlags`. And another function to check the status enabled/disabled of the specific feature flag which would be `fun isFeatureEnabled`. The rest could possibly be removed? 

`suspend fun fetchFeatureFlags` could simply return `Result<Unit>`. If I'm not missing something, what you care about there is whether refreshing the feature flags was successful or not. If successful then you know calling `fun isFeatureEnabled` will have fresh status. 

wdyt?

hafizrahman committed:
FluxC Bump

To make use of changes in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2734

hafizrahman committed:
Update the function to get a feature flag.

hafizrahman committed:
Refactor: Return `Return<Unit>` so the function simply checks whether fetching is successful or not.

hafizrahman committed:
Refactor: Simplify `RemoteFeatureFlag` as enum with no added functionalities.

hafizrahman committed:
Refactor: Put remote flag checking functionality as a use case class.

hafizrahman committed:
Add: comment regarding usage of `appVersion`

hafizrahman committed:
Add: Unit tests for WPComRemoteFeatureFlagRepository

hafizrahman committed:
Detekt fix.

hafizrahman committed:
Refactor: usecase class name to match the VerbNoun pattern.

hafizrahman committed:
Merge branch 'trunk' into issue/8998-internal-remote-feature-flag

hafizrahman commented:
@0nko @JorgeMucientes I've updated the code based on your various feedbacks, thanks again! Can you please have a look? I have also updated the main PR description with how things should work now.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/CheckRemoteFeatureFlagEnabled.kt:
```diff
@@ -0,0 +1,23 @@
+import com.woocommerce.android.util.RemoteFeatureFlag.LOCAL_NOTIFICATION_STORE_CREATION_READY
+import javax.inject.Inject
+
+class CheckRemoteFeatureFlagEnabled @Inject constructor(
```
Created a usecase class here so that:

1. `WPComRemoteFeatureFlagRepository` is now injected here, and this usecase is injectable to any classes/viewmodels that need it.
2. It becomes a place to allow force-enabling remote feature flag via `PackageUtils.isDebugBuild()`, which I feel better to be here instead of in the repository. This also mimics the way force-enabling is done in `FeatureFlag`.
3. Generally add better separation of concern.

hafizrahman replied:
The class name is a bit clunky imo ("CheckRemoteFeatureFlagEnabled") but it's the suggested VerbNoun pattern for a usecase class.

When in use, I'd probably suggest giving it a nicer name, e.g: `private val isRemoteFeatureFlagEnabled: CheckRemoteFeatureFlagEnabled`, so that the usage sounds better:

 `isRemoteFeatureFlagEnabled(FEATURE_FLAG)`

JorgeMucientes replied:
Thanks for applying the refactor @hafizrahman. The changes look good!
I also think it's perfectly fine if you name the use case class `IsRemoteFeatureFlagEnabled`. So feel free to name it that way if you prefer.

hafizrahman replied:
Thanks for having a look Jorge! I'll rename it then.

hafizrahman committed:
Merge branch 'trunk' into issue/8998-internal-remote-feature-flag

hafizrahman committed:
Refactor: rename to a clearer  IsRemoteFeatureFlagEnabled name.
